### PR TITLE
Added authentication

### PR DESCRIPTION
**Pull request by nb9960**

### CODE REVIEW

File: hackalog/settings.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import firebase_admin
 import os
 import pyAesCrypt
+from decouple import config
```
Comment (nishantwrp) [inline hackalog/settings.py]: You've used `decouple` here but it is not added in `Pipfile`. Make sure to use `pipenv` for managing dependencies.

File: hackalog/settings.py [hunk 2]
```diff
@@ -23,7 +24,7 @@
 # See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = os.environ.get('DJANGO_SECRET_KEY')
+SECRET_KEY = config('DJANGO_SECRET_KEY')
```
Comment (nishantwrp) [inline hackalog/settings.py]: Why this change?

File: authentication/admin.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+class ProfileInline(admin.StackedInline):
+  model = UserProfile
+  can_delete=False
+
+class UserAdmin(BaseUserAdmin):
+  inlines = (ProfileInline,)
+
+admin.site.unregister(User)
+admin.site.register(User, UserAdmin)
```
Comment (nishantwrp) [inline authentication/admin.py]: Add a new line at the EOF

File: authentication/serializers.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+class LoginSerializer(serializers.Serializer):
+    id_token = serializers.CharField(max_length=2400)
+
+    def validate_access_token(self, access_token):
+        return FirebaseAPI.verify_id_token(access_token)
+
+    def validate(self, data):
+        id_token = data.get('id_token', None)
+        user = None
+        jwt = self.validate_access_token(id_token)
```
Comment (nishantwrp) [inline authentication/serializers.py:19]: Maybe add a try-except block here to avoid any `500` errors. What say @krashish8?

File: authentication/serializers.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        uid = jwt['uid']
+        # pylint: disable=no-member
```
Comment (nishantwrp) [inline authentication/serializers.py]: Why pylint comments? There is no lint ci

File: authentication/serializers.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            user = User()
+            user.username = jwt['uid']
+            user.email = email
+            user.save()
+            user = user
+            profile = UserProfile.objects.create(
+                uid=uid, user=user, name=name, email=email)
+
+        data['user'] = user
+        return data
```
Comment (nishantwrp) [inline authentication/serializers.py]: new line.

File: authentication/serializers.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        if profile:
+            user = profile[0].user
+        else:
+            email = jwt['email']
+            name = jwt['name']
+            user = User()
+            user.username = jwt['uid']
+            user.email = email
+            user.save()
+            user = user
```
Comment (nishantwrp) [inline authentication/serializers.py]: Why this?

File: authentication/serializers.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            profile = UserProfile.objects.create(
+                uid=uid, user=user, name=name, email=email)
```
Comment (nishantwrp) [inline authentication/serializers.py:34]: Why store uid again in UserProfile when it is stored as an username.

File: authentication/urls.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from django.urls import path
+from .views import LoginView
+
+urlpatterns = [
+    path('login/', LoginView.as_view(), name='login'),
+]
```
Comment (nishantwrp) [inline authentication/urls.py]: new line

File: authentication/models.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+class UserProfile(models.Model):
+  uid = models.CharField(max_length=64, primary_key=True, editable=False)
+  user = models.OneToOneField(User, on_delete=models.CASCADE)
+  name=models.CharField(max_length=255)
+  email=models.EmailField(max_length=255)
+  # github_handle=models.CharField(max_length=255)
+
+  def __str__(self):
+    return self.name
```
Comment (nishantwrp) [inline authentication/models.py]: new line

File: authentication/serializers.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        jwt = self.validate_access_token(id_token)
+        uid = jwt['uid']
+        # pylint: disable=no-member
+        profile = UserProfile.objects.filter(uid=uid)
+
+        if profile:
+            user = profile[0].user
+        else:
+            email = jwt['email']
+            name = jwt['name']
```
Comment (nishantwrp) [inline authentication/serializers.py:27]: Remove this

**Review by nishantwrp — CHANGES_REQUESTED** [review]
Some comments ptal!

File: authentication/serializers.py (hunk 4 above)
Comment (krashish8) [reply authentication/serializers.py:19]: @nishantwrp I don't think it is necessary to add a try-except block here. This line will call the function `validate_access_token()` defined above, which has a try-except block, so that function will raise `ValidationError` if there is any `500` error from Firebase.

File: authentication/serializers.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    token = serializers.CharField(max_length=500)
+
+class LoginSerializer(serializers.Serializer):
+    id_token = serializers.CharField(max_length=2400)
+
+    def validate_access_token(self, access_token):
+        try:
+            return FirebaseAPI.verify_id_token(access_token)
+        except:
+            return serializers.ValidationError("Invalid Firebase Token")
```
Comment (krashish8) [inline authentication/serializers.py]:
```suggestion
            raise serializers.ValidationError("Invalid Firebase Token")
```

File: authentication/utils.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+from rest_framework.exceptions import ValidationError
+
+class FirebaseAPI:
+    @classmethod
+    def verify_id_token(cls, id_token):
+        try:
+            decoded_token = auth.verify_id_token(id_token)
+            return decoded_token
+        except ValueError:
+            raise ValidationError('Invalid Firebase ID Token.', HTTP_422_UNPROCESSABLE_ENTITY)
```
Comment (nishantwrp) [inline authentication/utils.py:12]: We have a try-except block here and in the serializers.py file as well. Can't we just do it once?

File: authentication/serializers.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        bio = validated_data['bio']
+        interests = validated_data['interests']
+        instance.name = name
+        instance.college = college
+        instance.github_handle = github_handle
+        instance.bio = bio
+        instance.interests = interests
+        instance.save()
+        return instance
+
```
Comment (nishantwrp) [inline authentication/serializers.py]: I don't think we need it since we are using `RetrieveUpdateAPIView`.

/cc @krashish8

File: authentication/models.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+class UserProfile(models.Model):
+  uid = models.CharField(max_length=64, primary_key=True, editable=False)
+  user = models.OneToOneField(User, on_delete=models.CASCADE)
+  name=models.CharField(max_length=255)
+  email=models.EmailField(max_length=255)
+  college = models.CharField(max_length=255)
+  github_handle = models.CharField(max_length=255,blank=True)
+  bio = models.CharField(max_length=255,blank=True)
+  interests = models.CharField(max_length=255,blank=True)
```
Comment (nishantwrp) [inline authentication/models.py]: These both could be `TextField`?

File: authentication/models.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
+from django.db import models
+from django.contrib.auth.models import User
+# Create your models here.
+
+class UserProfile(models.Model):
+  uid = models.CharField(max_length=64, primary_key=True, editable=False)
```
Comment (nishantwrp) [inline authentication/models.py:6]: Uid should be stored as the username, right? So we don't need it.

File: authentication/models.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  user = models.OneToOneField(User, on_delete=models.CASCADE)
+  name=models.CharField(max_length=255)
+  email=models.EmailField(max_length=255)
```
Comment (nishantwrp) [inline authentication/models.py]: I saw the email is stored in the original `user` model. So, we can eliminate this field too?

File: authentication/models.py [hunk 18]
```diff
@@ -0,0 +1,17 @@
+from django.db import models
+from django.contrib.auth.models import User
+# Create your models here.
+
+class UserProfile(models.Model):
+  uid = models.CharField(max_length=64, primary_key=True, editable=False)
+  user = models.OneToOneField(User, on_delete=models.CASCADE)
+  name=models.CharField(max_length=255)
```
Comment (nishantwrp) [inline authentication/models.py:8]: Also, lets add a `handle` field.

**Review by nishantwrp — CHANGES_REQUESTED** [review]
Some comments, ptal!

- On calling the /login endpoint. A user with an empty username gets created. Shouldn't it be the uid that is given by firebase?

File: authentication/serializers.py (hunk 14 above)
Comment (krashish8) [reply authentication/serializers.py]: Yes, I agree.

File: authentication/models.py (hunk 16 above)
Comment (nb9960) [reply authentication/models.py:6]: If we remove uid, it would be required to add id field and populate it. Also as you have mentioned to add handle field, so what's the need to have username field? Shouldn't we keep it blank?
